5126 Added Support for Case Only RECAP Alerts

albertisfu opened the pull request:
This PR fixes #5126 and works as follows:

A new option is shown in the Alerts form for RECAP alerts:

![Screenshot 2025-05-16 at 10 24 25 a m](https://github.com/user-attachments/assets/81551525-3746-4158-961f-259ef67c980b)

In other search types, such as Case Law or Oral Arguments, the regular form is displayed:

![Screenshot 2025-05-16 at 10 25 26 a m](https://github.com/user-attachments/assets/68d22ca2-468f-4126-8eb6-cbbed272b28c)

As talked about, it made sense to use the type `"d"` for case-only alerts, since we don’t expect to support a `"d"` alert type that behaves differently. Therefore, for RECAP search alerts, if the user selects:

- **Notifications on Cases only**: the alert is created with `alert_type = SEARCH_TYPES.DOCKETS` (`"d"`)
- **Notifications on both Cases and Filings**: the alert is created with the regular `alert_type = SEARCH_TYPES.RECAP` (`"r"`)

Default option is: `Notifications on both Cases and Filings`

The `alert_type` is stored in the `Alert` model. The search type in the query string remains unchanged. For example, for a type `"r"` alert:

`q=testing&type=r&order_by=score%20desc`

This means we don’t need to modify the query logic, as alerts continue to use the `"r"` type under the hood.

In the user Alerts panel, case-only alerts are labeled as `RECAP Cases only`:

![Screenshot 2025-05-16 at 10 39 34 a m](https://github.com/user-attachments/assets/9a61bb85-68ae-4ad9-9e09-a01a6794b321)

### Case-only Alerts Filtering Logic

- **Notifications on both Cases and Filings** behave exactly as before.
- **Notifications on Cases only** include the following change:

The triggering logic remains the same, the alert can be triggered by either a Docket or a RECAPDocument. The only addition is a new Redis set per alert:

`alert_hits:{alert_id}:co`

Once the alert is triggered by either a Docket or a RECAPDocument, the associated `docket_id` is added to this set. When sending alerts, it first checks if it's a case-only alert. If so, it verifies whether the matched `docket_id` is already in the `"co"` set. If it is, the alert is skipped; otherwise, it is triggered. This ensures that a case-only alert is only triggered **once** either by a Docket or a RECAPDocument.

This logic applies to both the Sweep index and the Percolator approach.

### Estimation Count for RECAP Case-only Alerts
- For **Notifications on both Cases and Filings**, the estimation count remains unchanged: Dockets matched with `date_filed` in the last 100 days + Filings matched with `entry_date_filed` in the last 100 days`
    
- For **Notifications on Cases only**, the estimation logic is different because a case-only query should trigger only once per case, regardless of whether it was triggered by the Docket or one of its RECAPDocuments.

So the estimation strategy is:
- `parent_hits` = Dockets that matched with `date_filed` within the last 100 days
- `child_hits` = Filings that matched with `entry_date_filed` within the last 100 days, **grouped by `docket_id`** (one filing per case)

Then this evaluation is used:
`total_recap_case_only_estimation = max(parent_hits, child_hits)`

Using max() makes sense, as the case-only alert will trigger once, whichever comes first.

The structure of the triggered alert emails and webhooks remains unchanged for case-only alerts. If the alert is triggered by one or more filings, the nested hits are still displayed as usual:
![Screenshot 2025-05-16 at 11 05 52 a m](https://github.com/user-attachments/assets/42a36701-6b95-4c64-b73c-d30d2d92aca6)

One question here is: **Should we visually distinguish case-only alerts from regular alerts in the email template?**

Finally. We need to update the RECAP Search Alerts documentation to include the latest information on case-only alerts. Additionally, since alerts can also be created via the API, we must document that case-only alerts can be created by setting` alert_type = DOCKETS` in the request, to avoid confusion. I'll open an issue for this.

mlissner commented:
This is great. A few things off the bat:

1. The alert count estimations should float to the right as in the spec.

3. This feels wrong to me: 

         total_recap_case_only_estimation = max(parent_hits, child_hits)

    Isn't the number of hits the number of dockets (i.e., the parent_hits)? If there are 1,000 documents that match, but they're all from a single docket, the total number of case only hits is still 1, right, not `max(1, 1000)`? 

4. Regarding this:

    > One question here is: Should we visually distinguish case-only alerts from regular alerts in the email template?

    I think it's fine not to distinguish in the alert. 

Thank you!

albertisfu commented:
> This feels wrong to me:
> 
>   total_recap_case_only_estimation = max(parent_hits, child_hits)
> Isn't the number of hits the number of dockets (i.e., the parent_hits)? If there are 1,000 documents that match, but they're all from a single docket, the total number of case only hits is still 1, right, not max(1, 1000)?

Well, for case-only alerts, `child_hits` refers to filings that matched the `entry_date_filed` within the last 100 days, grouped by `docket_id` (one filing per case).

This is a cardinality aggregation based on `docket_id`, so if a single case has 1,000 documents, the result from this query would be 1, meaning: `max(1, 1) = 1` in your example.

However, I think the `max` value is still useful because it can properly account for cases like the following:

**Docket 1** (filed within the last 100 days)
– Has 2 filings in the last 100 days

**Docket 2** (filed 105 days ago — outside the range)
– Has 5 filings in the last 100 days

So in this scenario, we would have:

- `parent_hits`: 1
- `child_hits`: 2 (one per each case)

Which gives us: `max(1, 2)`

In this case, the alert had **2 hits** in the last 100 days.

Does that make sense?

mlissner commented:
> Does that make sense?

Yep! Thanks for explaining.

mlissner reviewed (approved):
Code looks about right at a skim. I'll let @ERosendo give it a full look!

mlissner commented:
Alberto, I set Eduardo as reviewer, but it's still assigned to you until you get the right alignment bug tweaked, so please assign him once you've got it. Thank you both!

albertisfu commented:
Sure. I've fixed the right alignment of count estimates.

![Screenshot 2025-05-16 at 4 16 37 p m](https://github.com/user-attachments/assets/25b9f777-cbe0-43a1-9250-d5b993ad770b)

I also fixed other bugs I noticed when saving alerts of type “o” and “oa,” and ensured that the RECAP radio buttons display correctly when the form is re-rendered after an error.

ERosendo commented on cl/alerts/forms.py:
```diff
@@ -6,13 +6,45 @@
+
+        # Hide the alert_type field for search types other than RECAP.
+        if current_type not in (SEARCH_TYPES.RECAP, SEARCH_TYPES.DOCKETS):
+            self.fields["alert_type"].widget = HiddenInput()
+            # Restore the original valid alert_type field choices, since
+            # they were overridden for RECAP
+            original_field = self._meta.model._meta.get_field("alert_type")
+            field = self.fields["alert_type"]
+            field.choices = original_field.choices
```
It looks like the `choices` attribute for the `alert_type` field in the `Alert` model is linked to the value of `SEARCH_TYPES.SUPPORTED_ALERT_TYPES`. So, we could try something like this:

```suggestion
            self.fields["alert_type"].widget = HiddenInput()
            self.fields["alert_type"].choices =  SEARCH_TYPES.SUPPORTED_ALERT_TYPES
```

albertisfu replied:
Thanks. Yeah this is simpler!

ERosendo reviewed (approved):
LGTM 🚀 ! Just one small comment to address before we merge.

ERosendo commented:
Sorry @albertisfu, I just spotted another alignment issue. The "Create alert" button in the alert modal should always be left-aligned according to issue #5126. However, when creating recap alerts, it's center-aligned, and it gets worse when editing an alert because the "delete alert" button is also shown.

Here's the desired alignment:

![image](https://github.com/user-attachments/assets/92890043-b282-46b5-97ec-c83dc2a1eaa6)


And here's how it currently looks:

![image](https://github.com/user-attachments/assets/39e68aa0-78d2-4332-8856-3c4476f0fa8e)

albertisfu commented:
@ERosendo Thanks for your suggestions.
I've already addressed the alignment issue with the Alerts buttons.

![Screenshot 2025-05-21 at 5 41 05 p m](https://github.com/user-attachments/assets/cdb7bea5-8341-41bb-9aa8-04fa602c9742)
![Screenshot 2025-05-21 at 5 40 51 p m](https://github.com/user-attachments/assets/fe798748-03a4-438f-bbfe-5f5f9bce1cfb)

Let me know what do you think.

ERosendo reviewed (approved):
Thanks @albertisfu. LGTM :shipit:
